### ISSUE TITLE
Processing stuck after 8 minutes

### ISSUE THREAD

**Issue by dhafley** [pixee/codemodder-java#270]
When running on Apache Fink, the process hangs after eight minutes. It appears to hang
[codemodder-java-jstack-output.txt](https://github.com/pixee/codemodder-java/files/13743697/codemodder-java-jstack-output.txt)
 running codemod: pixee:java/prevent-filewriter-leak-with-nio. Attached is the output from jstack

**Comment by nahsra** [comment]
Can you capture 2 back to back `jstack` invocations when its hung? That may help us quickly understand if we have a deadlock.

**Comment by nahsra** [comment]
Feel free to re-submit with more logs!

**Comment by nahsra** [comment]
Re-opening this to comment after we saw similar behavior with Apache Pinot with similar logs to what we saw before. Keeping some notes here. 

The hangup deterministically occurs during `pixee:java/make-prng-seed-unpredictable` codemod execution from the [pixee CLI](https://github.com/pixee/pixee-cli/).

Running just this codemod on its own, it doesn't seem to get stuck:
```
./gradlew :core-codemods:run --args="--verbose --dry-run --codemod-include=pixee:java/make-prng-seed-unpredictable /tmp/code/pinot/"
```
This execution takes a minute.

The error also does not occur when I run all codemods against pinot:
```
./gradlew :core-codemods:run --args="--verbose --dry-run --codemod-include=pixee:java/make-prng-seed-unpredictable /tmp/code/pinot/"
```
This execution takes 5 minutes.

I ran both scenarios multiple times. This is an example thread dump seen during 

```
"main" #1 prio=5 os_prio=31 cpu=68362407.83ms elapsed=69739.14s tid=0x00007f946380e600 nid=0x2803 runnable  [0x000070000a46c000]
   java.lang.Thread.State: RUNNABLE
	at io.codemodder.DefaultCodemodExecutor.execute(DefaultCodemodExecutor.java:171)
	at io.codemodder.CLI.call(CLI.java:424)
	at io.codemodder.CLI.call(CLI.java:44)
...
 "pool-22-thread-1" #149 prio=5 os_prio=31 cpu=21450.33ms elapsed=69511.79s tid=0x00007f9464e64200 nid=0x942b runnable  [0x000070000b7a9000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileOutputStream.writeBytes(java.base@17.0.9/Native Method)
	at java.io.FileOutputStream.write(java.base@17.0.9/FileOutputStream.java:349)
	at java.io.BufferedOutputStream.flushBuffer(java.base@17.0.9/BufferedOutputStream.java:81)
	at java.io.BufferedOutputStream.flush(java.base@17.0.9/BufferedOutputStream.java:142)
	- locked <0x000000070020c038> (a java.io.BufferedOutputStream)
	at java.io.PrintStream.write(java.base@17.0.9/PrintStream.java:570)
	- locked <0x000000070020c010> (a java.io.PrintStream)
	at sun.nio.cs.StreamEncoder.writeBytes(java.base@17.0.9/StreamEncoder.java:234)
	at sun.nio.cs.StreamEncoder.implFlushBuffer(java.base@17.0.9/StreamEncoder.java:313)
	at sun.nio.cs.StreamEncoder.flushBuffer(java.base@17.0.9/StreamEncoder.java:111)
	- locked <0x000000070020c188> (a java.io.OutputStreamWriter)
	at java.io.OutputStreamWriter.flushBuffer(java.base@17.0.9/OutputStreamWriter.java:178)
	at java.io.PrintStream.writeln(java.base@17.0.9/PrintStream.java:723)
	- locked <0x000000070020c010> (a java.io.PrintStream)
	at java.io.PrintStream.println(java.base@17.0.9/PrintStream.java:1051)
	at java.lang.Throwable$WrappedPrintStream.println(java.base@17.0.9/Throwable.java:763)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:673)
	- locked <0x000000070020c010> (a java.io.PrintStream)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:659)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:650)
	at io.codemodder.DefaultCodemodExecutor.lambda$execute$0(DefaultCodemodExecutor.java:157)
	at io.codemodder.DefaultCodemodExecutor$$Lambda$1049/0x0000000134540000.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@17.0.9/Executors.java:539)
```

And from another run:

```
"main" #1 prio=5 os_prio=31 cpu=6815.98ms elapsed=384.17s tid=0x00007ff157808a00 nid=0x2303 waiting on condition  [0x0000700007e5c000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@17.0.9/Native Method)
	- parking to wait for  <0x000000070a743200> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@17.0.9/LockSupport.java:252)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(java.base@17.0.9/AbstractQueuedSynchronizer.java:1672)
	at java.util.concurrent.ThreadPoolExecutor.awaitTermination(java.base@17.0.9/ThreadPoolExecutor.java:1464)
	at io.codemodder.DefaultCodemodExecutor.execute(DefaultCodemodExecutor.java:164)
	at io.codemodder.CLI.call(CLI.java:424)
...
"pool-22-thread-1" #149 prio=5 os_prio=31 cpu=16416.59ms elapsed=185.04s tid=0x00007ff15a008c00 nid=0xb25f runnable  [0x000070000939e000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileOutputStream.writeBytes(java.base@17.0.9/Native Method)
	at java.io.FileOutputStream.write(java.base@17.0.9/FileOutputStream.java:349)
	at java.io.BufferedOutputStream.flushBuffer(java.base@17.0.9/BufferedOutputStream.java:81)
	at java.io.BufferedOutputStream.flush(java.base@17.0.9/BufferedOutputStream.java:142)
	- locked <0x00000007001de008> (a java.io.BufferedOutputStream)
	at java.io.PrintStream.write(java.base@17.0.9/PrintStream.java:570)
	- locked <0x00000007001ddfe0> (a java.io.PrintStream)
	at sun.nio.cs.StreamEncoder.writeBytes(java.base@17.0.9/StreamEncoder.java:234)
	at sun.nio.cs.StreamEncoder.implFlushBuffer(java.base@17.0.9/StreamEncoder.java:313)
	at sun.nio.cs.StreamEncoder.flushBuffer(java.base@17.0.9/StreamEncoder.java:111)
	- locked <0x00000007001de100> (a java.io.OutputStreamWriter)
	at java.io.OutputStreamWriter.flushBuffer(java.base@17.0.9/OutputStreamWriter.java:178)
	at java.io.PrintStream.writeln(java.base@17.0.9/PrintStream.java:723)
	- locked <0x00000007001ddfe0> (a java.io.PrintStream)
	at java.io.PrintStream.println(java.base@17.0.9/PrintStream.java:1051)
	at java.lang.Throwable$WrappedPrintStream.println(java.base@17.0.9/Throwable.java:763)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:673)
	- locked <0x00000007001ddfe0> (a java.io.PrintStream)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:659)
	at java.lang.Throwable.printStackTrace(java.base@17.0.9/Throwable.java:650)
	at io.codemodder.DefaultCodemodExecutor.lambda$execute$0(DefaultCodemodExecutor.java:157)
	at io.codemodder.DefaultCodemodExecutor$$Lambda$1065/0x00000001355406b0.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@17.0.9/Executors.java:539)
```

Now, each codemod has a timeout of 15 minutes:
```java
boolean success = executor.awaitTermination(15, TimeUnit.MINUTES);
log.trace("Success running codemod: {}", success);
while (!executor.isTerminated()) {
  final Future<String> future = service.poll();
  if (future != null) {
    log.trace("Finished: {}", future.get());
  }
}
```

After the 15 minute timeout, the stack changes a little bit:
```
"main" #1 prio=5 os_prio=31 cpu=1176579.07ms elapsed=2271.14s tid=0x00007ff157808a00 nid=0x2303 runnable  [0x0000700007e5b000]
   java.lang.Thread.State: RUNNABLE
	at io.codemodder.DefaultCodemodExecutor.execute(DefaultCodemodExecutor.java:171)
	at io.codemodder.CLI.call(CLI.java:424)
	at io.codemodder.CLI.call(CLI.java:44)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
```

This suggests that the `awaitTermination()` call finishes (the timeout occurs) and one of these calls is hanging:
```
162:    executor.shutdown();
163:    try {
164:      boolean success = executor.awaitTermination(15, TimeUnit.MINUTES);
165:      log.trace("Success running codemod: {}", success);
166:      while (!executor.isTerminated()) {
167:        final Future<String> future = service.poll();
168:        if (future != null) {
169:          log.trace("Finished: {}", future.get());
```

There is some unfortunate nonsense here in that the line numbers don't match the 0.69.2 tag contents (line 171 doesn't contain any code.) I am going to ask the CLI team to re-publish with an updated artifact and test more.

**Comment by nahsra** [comment]
This line was causing problems:
```java
e.printStackTrace();
```

This caused a hang, and it is still not clear why -- but every hung run included another thread that was inside of it. I read about other people having issues with threads concurrently printing stack traces and having issues with locks, so I just experimented with pushing this problem to the logger in the hopes it has more built-in smarts for avoiding issues like this:

```diff
- e.printStackTrace();
+ log.error("Problem scanning file", e);
```

This, for some reason, does the trick. I don't have the luxury time of figuring out why. I'll push a new release. I also lowered the "timeout" for a particular codemod from 15 minutes to 5.
